### PR TITLE
[NOT YET] Large TSV Downloads

### PR DESCRIPTION
**Pull request by rushtong**
NOT READY YET!
Addresses https://broadinstitute.atlassian.net/browse/GAWB-1998

### Changes in this PR:
* Added a streaming actor that does not use the PerRequest pattern
* Use paginated endpoint from Rawls to get all entities
* Added a async capabilities to the HttpRawlsDAO

### Testing:
* Generate a workspace with lots of entities. I have a groovy script that does this - touch base with me and we can make some participant files of any size we need. 
* Export entities from data tab.

---

- [x] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Make sure Swagger is updated if API changes
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell the tech lead (TL) that the PR exists if they wants to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Product Owner** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpRawlsDAO.scala [hunk 1]
```diff
@@ -26,6 +30,21 @@ import scala.concurrent.{ExecutionContext, Future}
 class HttpRawlsDAO( implicit val system: ActorSystem, implicit val executionContext: ExecutionContext )
   extends RawlsDAO with RestJsonClient {
 
+  // TODO: Copied from HttpOntologyDAO. How to refactor to a common trait ???
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/HttpRawlsDAO.scala]: Maybe put it in `RestJsonClient`?
```
private def getHostConnectorSetup(address, port, scheme): HostConnectorSetup = HostConnectorSetup(...)

protected final def getHostConnector(address, port, scheme): Future[ActorRef] = {
  implicit val timeout:Timeout = 60.seconds // timeout to get the host connector reference
  for (Http.HostConnectorInfo(connector, _) <- IO(Http) ? getHostConnectorSetup(address, port, scheme)) yield connector
}
```
Just a thought. Then any subclass could potentially build their own connector for async operations.

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/ExportEntitiesByTypeActor.scala [hunk 2]
```diff
@@ -41,6 +46,63 @@ trait ExportEntitiesByType extends FireCloudRequestBuilding {
...
+  def streamEntities(ctx: RequestContext, workspaceNamespace: String, workspaceName: String, filename: String, entityType: String, attributeNames: Option[IndexedSeq[String]]): Future[Stream[String]] = {
+    val sortField = entityType + "_id"
+    val firstQuery: EntityQuery = EntityQuery(page = 1, pageSize = 1, sortField = sortField, sortDirection = SortDirections.Ascending, filterTerms = None)
+
+    // Get all of the possible queries
+    lazy val pageQueriesFuture: Future[Seq[EntityQuery]] = getQueryResponse(workspaceNamespace, workspaceName, entityType, firstQuery) map {
+      queryResponse =>
+        val pageSize = 500 // TODO: Should this be a config?
+      val pages = queryResponse.resultMetadata.filteredCount/pageSize match {
+        case x if x > Math.floor(x) => (Math.floor(x) + 1).toInt
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/ExportEntitiesByTypeActor.scala]: It looks like `queryResponse.resultMetadata.filteredCount` and `pageSize` are both `Int`s, so won't `x` be an `Int` already? So no need for `Math.floor` and casting via `toInt`.

Comment (rushtong) [reply src/main/scala/org/broadinstitute/dsde/firecloud/service/ExportEntitiesByTypeActor.scala]: Good catch. I thought the division would coerce it to a non-Int. I am happily wrong!

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/ExportEntitiesByTypeActor.scala [hunk 3]
```diff
@@ -41,6 +46,63 @@ trait ExportEntitiesByType extends FireCloudRequestBuilding {
...
+      }
+        val range = 1 to pages
+        range map {
+          page =>
+            EntityQuery(page = page, pageSize = pageSize, sortField = sortField, sortDirection = SortDirections.Ascending, filterTerms = None)
+        }
+    }
+
+    // Make those queries and generate a nested mess of streams
+    lazy val nestedEntityFutures: Future[Seq[Future[Seq[Entity]]]] = pageQueriesFuture map { querySeq =>
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/ExportEntitiesByTypeActor.scala]: This is where'd I'd use `Traverse`!

So you have a:
```
pageQueriesFuture: Future[Seq[EntityQuery]]
```
and you want to get a:
```
Future[Seq[Entity]]
```
So:
```
val entitiesFuture: Future[Seq[Entity]] = pageQueriesFuture.flatMap(_.traverse(q => getEntities(workspaceNamespace, workspaceName, entityType, q)).map(_.flatten))
```
Does that work? I'm compiling with my brain...

Comment (rushtong) [reply src/main/scala/org/broadinstitute/dsde/firecloud/service/ExportEntitiesByTypeActor.scala]: @rtitle - Yes, this whole section of the code needs a lot of work. I'm pretty sure it's not working the way I really want it to ... I'll likely ask for your help next week :-)

**Comment by rushtong** [comment]
Closing this PR. Will start a fresh branch with minimal changes necessary.

**Pull request closed without merging**